diff --git a/types/jsx.d.ts b/types/jsx.d.ts
index a5417ab5f9bbcbc92b5bc232798251d6cd1ab227..b7cf179d6363afd9368299c55ddd1ffe3682f07d 100644
--- a/types/jsx.d.ts
+++ b/types/jsx.d.ts
@@ -9,19 +9,24 @@ import * as csstype from 'csstype';
 type DOMElement = Element;
 
 export namespace JSX {
-  type Element =
-    | Node
-    | ArrayElement
-    | FunctionElement
-    | (string & {})
-    | number
-    | boolean
-    | null
-    | undefined;
+  type FunctionMaybe<T = unknown> = ({ (): T }) | T;
+  interface ElementRegistry {
+    "Node": Node
+    "ArrayElement": ArrayElement
+    "FunctionElement": FunctionElement
+    "(string & {})": (string & {})
+    "number": number
+    "boolean": boolean
+    "null": null
+    "undefined": undefined
+  }
+  
+  type Element = ElementRegistry[keyof ElementRegistry];
   interface ArrayElement extends Array<Element> {}
   interface FunctionElement {
     (): Element;
   }
+
   interface ElementClass {
     // empty, libs can define requirements downstream
   }